hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -87,26 +95,15 @@ public static CompletableFuture<Map<MemberInfo, ExecutionPlan>> createExecutionP
         final int defaultParallelism = nodeEngine.getConfig().getJetConfig().getCooperativeThreadCount();
         final EdgeConfig defaultEdgeConfig = nodeEngine.getConfig().getJetConfig().getDefaultEdgeConfig();
         final Set<Integer> requiredPartitions = jobConfig.getArgument(KEY_REQUIRED_PARTITIONS);
-        final boolean memberPruningUsed = requiredPartitions != null;
+        final EnumSet<PartitionPruningLevel> ppl = requiredPartitions != null
+                ? Objects.requireNonNull(jobConfig.getArgument(KEY_MEMBER_PRUNING_LEVEL))
```

Comment:
how about determining that via DAG analysis:
- mandatory members are the ones specified in `Edge.distributedTo` (this would handle COORDINATOR_REQUIRED case)
- if `Edge.partitioner instanceOf Single` then Single.key is extra partition needed to include in the job
- if any `Edge.partitioner != null` and is not `Single` then all partitions are extra partitions (`ALL_PARTITIONS_REQUIRED`)

extra partition = does not impact member choice, but is added to list of partitions in the job

Reply:
During implementing this I started with DAG analysis, Later, I realized that I don't want extra graph traversal: SQL is building DAG, so SQL also can provide feedback about prunability level. We have no intention to make it universal, so we won't burn additional cycles to traverse DAG :)

Reply:
how many cycles that are compared to those burned on SQL optimization? Are you sure you are optimizing a real bottleneck in that way? And we pay other costs of that

>We have no intention to make it universal

Actually, why not? Jet is universal, partition pruning also can be.

Reply:
Well, I see implemented approach as the best one. If we want to make it universal and dig deeper, we may see cases when DAG analysis may not be universal as we might imagine.

Reply:
Also, it is enough for a given scope.

Reply:
Analysis on Jet side implemented in 2c93a4ffd49834e928a8ecba7799818d7b1d24aa